pkg/database/truncator.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2022 Codenotary Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package database
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/codenotary/immudb/embedded/store"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	// ErrNoTxBeforeTime is returned when retention period is not reached.
+	ErrRetentionPeriodNotReached = errors.New("retention period not reached")
+)
+
+// Truncator provides truncation against an underlying storage
+// of appendable data.
+type Truncator interface {
+	// Plan returns the transaction upto which the log can be truncated.
+	// When resulting transaction before specified time does not exists
+	//  * No transaction header is returned.
+	//  * Returns nil TxHeader, and an error.
+	Plan(time.Time) (*store.TxHeader, error)
+
+	// Truncate runs truncation against the relevant appendable logs. Must
```

Comment:
there shouldn't be a strong dependency on `Plan`, calling truncation up to a transaction should be possible without calling `Plan`

Reply:
In terms of the truncator, plan is required to figure out the transaction header up to which truncation can be done, so it seems apt for this interface